config/experimental/auth/auth.go

```diff
@@ -0,0 +1,218 @@
+// Package auth is an internal package that provides authentication utilities.
+//
+// IMPORTANT: This package is not meant to be used directly by consumers of the
+// SDK and is subject to change without notice.
+package auth
+
+import (
+	"sync"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+const (
+	// Default duration for the stale period. The number as been set arbitrarily
+	// and might be changed in the future.
+	defaultStaleDuration = 3 * time.Minute
+
+	// Disable the asynchronous token refresh by default. This is meant to
+	// change in the future once the feature is stable.
+	defaultDisableAsyncRefresh = true
+)
+
+type Option func(*cachedTokenSource)
+
+// WithCachedToken sets the initial token to be used by a cached token source.
+func WithCachedToken(t *oauth2.Token) Option {
+	return func(cts *cachedTokenSource) {
+		cts.cachedToken = t
+	}
+}
+
+// WithStaleDuration sets the duration for which a token is considered stale.
+// Stale tokens are still valid but will trigger an asynchronous refresh if
+// async refresh is enabled. The default value is 3 minutes.
+func WithStaleDuration(d time.Duration) Option {
+	return func(cts *cachedTokenSource) {
+		cts.staleDuration = d
+	}
+}
+
+// WithAsyncRefresh enables or disables the asynchronous token refresh.
+func WithAsyncRefresh(b bool) Option {
+	return func(cts *cachedTokenSource) {
+		cts.disableAsync = !b
+	}
+}
+
+// NewCachedTokenProvider wraps a [oauth2.TokenSource] to cache the tokens
+// it returns. By default, the cache will refresh tokens asynchronously a few
+// minutes before they expire.
+//
+// The token cache is safe for concurrent use by multiple goroutines and will
+// guarantee that only one token refresh is triggered at a time.
+//
+// The token cache does not take care of retries in case the token source
+// returns and error; it is the responsibility of the provided token source to
+// handle retries appropriately.
+//
+// If the TokenSource is already a cached token source (obtained by calling this
+// function), it is returned as is.
+func NewCachedTokenSource(ts oauth2.TokenSource, opts ...Option) oauth2.TokenSource {
```

Comment:
Why return an interface rather than explicit class? This is all experimental so we don't need to worry about details changing. [Preemptive interfacing](https://bryanftan.medium.com/accept-interfaces-return-structs-in-go-d4cab29a301b) is often considered anti-pattern. Maybe you have a good reason - but it's no clear.

If you agree to this, then follow up question is why not export the class cachedTokenSource -> CachedTokenSource?

Once you exported the class, you can get rid of options and let users set StaleDuration and DisableAsync directly on the class. 

In other words, it would be simpler if this followed Cmd/Command pattern:

https://pkg.go.dev/os/exec#Cmd
https://pkg.go.dev/os/exec#Command

Reply:
If that's the pattern you're using everywhere in SDK then it makes sense to follow it here as well.